### PR TITLE
Support regression models in several inference attacks

### PR DESCRIPTION
**Pull request by GalHorowitz**
# Description

- New Scikitlearn regression wrapper, ScikitlearnDecistionTreeRegressor moved to `art.estimators.regression`
- Added a fixture for the diabetes regression dataset from scitkitlearn
- _estimator_requirements now supports tuples to represent unions
- MembershipInferenceBlackBox now supports regression models (when input_type='loss')
- AttributeInferenceBlackBox and MembershipInferenceBlackBox now support regression models
- Two new notebooks to demonstrate membership and attribute inference on regression models
- New attribute inference baseline that uses the true label, used for evaluating attribute inference attacks

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

All unit tests pass.

**Test Configuration**:
- OS: MacOS 11.4
- Python version: 3.8.11

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/inference/attribute_inference/black_box.py [hunk 1]
```diff
@@ -47,23 +48,28 @@ class AttributeInferenceBlackBox(AttributeInferenceAttack):
...
         """
         Create an AttributeInferenceBlackBox attack instance.
 
-        :param classifier: Target classifier.
+        :param model: Target model.
         :param attack_model: The attack model to train, optional. If none is provided, a default model will be created.
         :param attack_feature: The index of the feature to be attacked or a slice representing multiple indexes in
                                case of a one-hot encoded feature.
+        :param prediction_normal_factor: If supplied, predictions of the model are multiplied by the factor when
+                                         when used as inputs to the attack-model. Only applicable when `model` is a
```
Comment (abigailgold) [inline art/attacks/inference/attribute_inference/black_box.py:69]: double "when"

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    The idea is to train a simple neural network to learn the attacked feature from the rest of the features, and the
+    true label. Should be used to compare with other attribute inference results.
+    """
+
+    _estimator_requirements = ()
+
+    def __init__(
+        self,
+        attack_model: Optional["CLASSIFIER_TYPE"] = None,
```
Comment (abigailgold) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:51]: Why does this attack only support classifiers? Should be both (same as the regular black-box attack)

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                nesterovs_momentum=True,
+                early_stopping=False,
+                validation_fraction=0.1,
+                beta_1=0.9,
+                beta_2=0.999,
+                epsilon=1e-08,
+                n_iter_no_change=10,
+                max_fun=15000,
+            )
+        # self._scaler = StandardScaler()
```
Comment (abigailgold) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:98]: Need to remove these comments. Suggest to use normalizing factor for regression case, same as the black-box attack.

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        self._check_params()
+
+    def fit(self, x: np.ndarray, true_y: np.ndarray) -> None:
+        """
+        Train the attack model.
+
+        :param x: Input to training process. Includes all features used to train the original model.
+        :param true_y: True labels of the features.
```
Comment (abigailgold) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:106]: Suggest to call this "y" (to be consistent)

File: art/attacks/inference/membership_inference/black_box.py [hunk 5]
```diff
@@ -184,8 +192,6 @@ def fit(  # pylint: disable=W0613
             test_features = self.estimator.predict(test_x).astype(np.float32)
         # only for models with loss
         elif self.input_type == "loss":
-            if NeuralNetworkMixin not in type(self.estimator).__mro__:
```
Comment (abigailgold) [inline art/attacks/inference/membership_inference/black_box.py:187]: Shouldn't we check here for loss mixin?

Comment (abigailgold) [reply art/attacks/inference/membership_inference/black_box.py:187]: Never mind, saw you moved this to check_params

File: art/utils.py [hunk 6]
```diff
@@ -1028,7 +1061,8 @@ def load_dataset(
     name: str,
 ) -> DATASET_TYPE:
     """
-    Loads or downloads the dataset corresponding to `name`. Options are: `mnist`, `cifar10` and `stl10`.
+    Loads or downloads the dataset corresponding to `name`. Options are: `mnist`, `cifar10`, `stl10`, `iris`, `nursery`
+    and `stl10`.
```
Comment (abigailgold) [inline art/utils.py:1065]: stl10 appears twice. I think this should be diabetes

File: art/estimators/classification/scikitlearn.py [hunk 7]
```diff
@@ -35,6 +35,7 @@
     ClassGradientsMixin,
     ClassifierMixin,
 )
+from art.estimators.regression.scikitlearn import ScikitlearnDecisionTreeRegressor
```
Comment (abigailgold) [inline art/estimators/classification/scikitlearn.py:38]: @beat-buesser Are you sure that ScikitlearnDecisionTreeRegressor should be used within the get_trees method of ScikitlearnGradientBoostingClassifier?

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 2 above)
Comment (GalHorowitz) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:51]: The `attack_model` parameter is for the model that the baseline attack trains, it must be a classifier because its output is the predicted attribute

Comment (abigailgold) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:51]: Oh, right. Ok.

File: art/estimators/classification/scikitlearn.py (hunk 7 above)
Comment (beat-buesser) [reply art/estimators/classification/scikitlearn.py:38]: @abigailgold Yes, I think that should be correct. It looks like the trees of the ensemble in `sklearn.ensemble.GradientBoostingClassifier` are of type `DecisionTreeRegressor` in https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/ensemble/_gb.py#L195

**Review by beat-buesser — APPROVED** [review]
Hi @GalHorowitz Thank you very much for this great pull request adding support for regression models in inference attacks!
